### PR TITLE
ITT cycles

### PR DESCRIPTION
**Pull request by alaric-dfe**
Import data about ITT cycles from https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/services/cycle_timetable.rb in the Apply codebase.

See trello card https://trello.com/c/E2Su8JFh/876-put-recruitment-cycles-in-dfe-reference-data for details - in particular, they want them accessible via BigQuery.

Also includes adding some new types to the schema system and bigquery importer to support things found in this data!

### CODE REVIEW

File: lib/dfe/reference_data/itt/cycles.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      end
+
+      CYCLES = DfE::ReferenceData::HardcodedReferenceList.new(
+        {
+          2019 => {
+            find_opens: make_local_time(2018, 10, 6, 9),
+            apply_opens: make_local_time(2018, 10, 13, 9),
+            apply_1_deadline: make_local_time(2019, 8, 24, 18),
+            apply_2_deadline: make_local_time(2019, 9, 18, 18),
+            end_of_cycle: make_local_time(2021, 9, 29, 23, 59, 59),
```
Comment (frankieroberto) [inline lib/dfe/reference_data/itt/cycles.rb:35]: Typo here?

```suggestion
            end_of_cycle: make_local_time(2019, 9, 29, 23, 59, 59),
```

File: lib/dfe/reference_data/itt/cycles.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            find_closes: make_local_time(2019, 10, 3, 23, 59, 59),
+            holidays: {}
+          },
+          2020 => {
+            find_opens: make_local_time(2019, 10, 6, 9),
+            apply_opens: make_local_time(2019, 10, 13, 9),
+            show_deadline_banner: make_local_time(2020, 8, 1, 9),
+            show_summer_recruitment_banner: make_local_time(2020, 7, 1, 9),
```
Comment (frankieroberto) [inline lib/dfe/reference_data/itt/cycles.rb:43]: I’d be tempted to drop these, as they're really only used for Apply and aren't defined externally. (Also subject to change if we decide to show the banners earlier or later)

```suggestion
```

File: lib/dfe/reference_data/itt/cycles.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      # Timezone used to specify dates+times in this file
+      TIMEZONE = TZInfo::Timezone.get('Europe/London')
+
+      def self.make_local_time(*args)
+        TIMEZONE.local_to_utc(DateTime.new(*args))
+      end
+
+      CYCLES = DfE::ReferenceData::HardcodedReferenceList.new(
+        {
+          2019 => {
```
[the remark this answers is not in the document]
Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb:30]: Done!

File: lib/dfe/reference_data/itt/cycles.rb (hunk 1 above)
Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb:35]: Is there a typo here, then? https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/services/cycle_timetable.rb#L10

File: lib/dfe/reference_data/itt/cycles.rb (hunk 2 above)
Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb:43]: Done

File: lib/dfe/reference_data/itt/cycles.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    module ITT
+      CYCLES_SCHEMA = {
+        id: :integer,
+        find_opens: :datetime,
+        apply_opens: :datetime,
+        show_deadline_banner: { kind: :optional, schema: :datetime },
+        show_summer_recruitment_banner: { kind: :optional, schema: :datetime },
+        apply_1_deadline: :datetime,
+        apply_2_deadline: :datetime,
+        end_of_cycle: :datetime,
```
[the remark this answers is not in the document]
Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb:16]: It was called `reject_by_default` in the original, but @alexchristensenedu thought `end_of_cycle` was better - I have no dogs in this fight, but I'll gladly help the two of you find an agreement :-) Is `provider_decision_deadline` good with both of you?

File: lib/dfe/reference_data/itt/cycles.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            apply_2_deadline: make_local_time(2020, 9, 18, 18),
+            end_of_cycle: make_local_time(2021, 9, 29, 23, 59, 59),
+            find_closes: make_local_time(2020, 10, 3, 23, 59, 59),
+            holidays: {}
+          },
+          2021 => {
+            find_opens: make_local_time(2020, 10, 6, 9),
+            apply_opens: make_local_time(2020, 10, 13, 9),
+            show_deadline_banner: make_local_time(2021, 8, 1, 9),
+            show_summer_recruitment_banner: make_local_time(2020, 7, 1, 9),
```
Comment (alexchristensenedu) [inline lib/dfe/reference_data/itt/cycles.rb:54]: @alaric-dfe  @frankieroberto  `show_summer_recruitment_banner` for 2021, has a 2020 date

File: lib/dfe/reference_data/itt/cycles.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexchristensenedu) [reply lib/dfe/reference_data/itt/cycles.rb:16]: I'm happy for `provider_decision_deadline` but there needs to be description on the fields, so it's clear what this mean

File: lib/dfe/reference_data/itt/cycles.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            holidays: {}
+          },
+          2020 => {
+            find_opens: make_local_time(2019, 10, 6, 9),
+            apply_opens: make_local_time(2019, 10, 13, 9),
+            show_deadline_banner: make_local_time(2020, 8, 1, 9),
+            show_summer_recruitment_banner: make_local_time(2020, 7, 1, 9),
+            apply_1_deadline: make_local_time(2020, 8, 24, 18),
+            apply_2_deadline: make_local_time(2020, 9, 18, 18),
+            end_of_cycle: make_local_time(2021, 9, 29, 23, 59, 59),
```
Comment (frankieroberto) [inline lib/dfe/reference_data/itt/cycles.rb:46]: This one looks wrong too... 🤦

File: lib/dfe/reference_data/itt/cycles.rb (hunk 5 above)
Comment (frankieroberto) [reply lib/dfe/reference_data/itt/cycles.rb:54]: 🤦

Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb]: At Frankie's suggestion I've removed those dates as they're just internal to Apply operations - you don't want them for anything analytical, do you @alexchristensenedu ?

File: lib/dfe/reference_data/itt/cycles.rb (hunk 6 above)
Comment (alaric-dfe) [reply lib/dfe/reference_data/itt/cycles.rb]: Yep, got it

File: lib/dfe/reference_data/itt/cycles.rb (hunk 5 above)
Comment (alexchristensenedu) [reply lib/dfe/reference_data/itt/cycles.rb]: @alaric-dfe nope I don't need the banner dates

**Pull request merged**